marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -155,43 +167,75 @@ data instance StorableEvent UtxoHandle = UtxoEvent
   , ueChainPoint  :: !C.ChainPoint
   } deriving (Eq, Ord, Show, Generic)
 
-eventIsBefore :: StorableEvent UtxoHandle  -> C.ChainPoint -> Bool
-eventIsBefore (UtxoEvent _ _ (C.ChainPoint _ slot)) (C.ChainPoint _ slot') =  slot < slot'
+eventIsBefore :: C.ChainPoint -> StorableEvent UtxoHandle -> Bool
+eventIsBefore (C.ChainPoint slot' _) (UtxoEvent _ _ (C.ChainPoint slot _)) =  slot <= slot'
 eventIsBefore _ _                                                          = False
 
+-- | mappend, combine Unspent utxoEvents:
+--
 instance Semigroup (StorableEvent UtxoHandle) where
-  e@(UtxoEvent u i cp) <> (UtxoEvent u' i' cp') =
-    if cp == cp' then
-      UtxoEvent (Set.union u u') (Set.union i i') cp
-    else e -- do not combine events from different chain points
+  (UtxoEvent us is cp) <> (UtxoEvent us' is' cp') =
+    UtxoEvent utxos txins (max cp cp')
+    where
+      toTxIn :: Utxo -> C.TxIn
+      toTxIn u = C.TxIn (u ^.txId) (u ^. txIx)
+      txins = Set.union is is'
+      utxos
+        = foldl' (\a c -> if toTxIn c `Set.notMember` txins then Set.insert c a;else a) Set.empty
+        . Set.union us
+        $ us'
```

Comment:
Minor detail. I know we don't have a coding style guide, but I suggest doing something like `foldl' ... $ Set.union us us'` instead of using function composition, because that is how we usually write it in the plutus-apps codebase.